core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -0,0 +1,610 @@
+     * @return true if the provided array maps to all indices, false otherwise
+     */
+    static boolean isAllIndices(List<String> aliasesOrIndices) {
+        return aliasesOrIndices == null || aliasesOrIndices.isEmpty() || isExplicitAllPattern(aliasesOrIndices);
+    }
+
+    /**
+     * Identifies whether the array containing index names given as argument explicitly refers to all indices
+     * The empty or null array doesn't explicitly map to all indices
+     *
+     * @param aliasesOrIndices the array containing index names
+     * @return true if the provided array explicitly maps to all indices, false otherwise
+     */
+    static boolean isExplicitAllPattern(List<String> aliasesOrIndices) {
+        return aliasesOrIndices != null && aliasesOrIndices.size() == 1 && MetaData.ALL.equals(aliasesOrIndices.get(0));
+    }
+
+    /**
+     * Identifies whether the first argument (an array containing index names) is a pattern that matches all indices
+     *
+     * @param indicesOrAliases the array containing index names
+     * @param concreteIndices  array containing the concrete indices that the first argument refers to
+     * @return true if the first argument is a pattern that maps to all available indices, false otherwise
+     */
+    boolean isPatternMatchingAllIndices(MetaData metaData, String[] indicesOrAliases, String[] concreteIndices) {
+        // if we end up matching on all indices, check, if its a wildcard parameter, or a "-something" structure
+        if (concreteIndices.length == metaData.concreteAllIndices().length && indicesOrAliases.length > 0) {
+
+            //we might have something like /-test1,+test1 that would identify all indices
+            //or something like /-test1 with test1 index missing and IndicesOptions.lenient()
+            if (indicesOrAliases[0].charAt(0) == '-') {
+                return true;
+            }
+
+            //otherwise we check if there's any simple regex
+            for (String indexOrAlias : indicesOrAliases) {
+                if (Regex.isSimpleMatchPattern(indexOrAlias)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static class Context {
+
+        private final ClusterState state;
+        private final IndicesOptions options;
+
+        public Context(ClusterState state, IndicesOptions options) {
+            this.state = state;
+            this.options = options;
+        }
+
+        public ClusterState getState() {
+            return state;
+        }
+
+        public IndicesOptions getOptions() {
+            return options;
+        }
+    }
+
+    private interface IndexExpressionFilter {
+
+        /**
+         * Expands or changes the list of expressions in concrete indices and aliases
+         *
+         * @return a new list with expressions based on the provided expressions
+         */
+        List<String> operate(Context context, List<String> expressions);
```

Comment:
not a big fan of calling it `operate`. If it's a filter maybe the method should be called `filter`. That said when I look at it I don't know anymore if it's really a filter, more of a resolver?